---
title: "KJOH - proj5: pH effect"
author: "Pieter van Veelen"
date: "2024-05-22"
output: html_document
---

### Data microbiome profiling data analysis of PHA accumulating biomass 

```{r , eval=TRUE, echo=T, include=T, message=FALSE, warning=FALSE}

# set optional parameters
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      message = FALSE)
options(scipen = 999, digits = 3)

```

```{r library loading, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

## load required packages
library(Hmisc)
library(phyloseq)
library(qiime2R)
library(tidyverse)
library(magrittr)
library(devtools)
library(qiime2R)
library(here)
library(breakaway)
library(DivNet)
library(openxlsx)
library(ape)
library(vegan)
library(ggtext)
library(cowplot)
library(RColorBrewer)
library(microbiome)
library(lme4)
library(lmerTest)
library(decontam)
library(ampvis2)
library(glue)
library(lubridate)
library(DECIPHER)
library(ensembleTax)
library(mEQO)
library(ggh4x)
library(AICcPermanova)

```

```{r project organization, message=F, echo=T, eval=T, warning=F, include=T, cache=T}

# project name
proj = "KJOH_proj5_Q17636"

# create directories
if(!dir.exists("figures")){dir.create("figures")}
if(!dir.exists("output_data")){dir.create("output_data")} 
if(!dir.exists("scripts")){dir.create("scripts")} 
#if(!dir.exists("scripts/QIIME2")){dir.create("scripts/QIIME2")} 

### create color vector for later use
library(RColorBrewer)

colorset <- c("darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen",  "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey", "darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen", "deeppink", "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "darkblue", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey")


# define a color vector using brewer.pal and colorset
colvec = c(brewer.pal(8,"Dark2"),brewer.pal(8,"Set2"),brewer.pal(8,"Accent"),brewer.pal(8,"Paired"),brewer.pal(12,"Set3"),brewer.pal(9,"Set1"), brewer.pal(9,"YlGnBu"), colorset)


```


```{r import data, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

physeq = 
qza_to_phyloseq(
  features = "input_data/KJOH_16S_515F_926R_08052024_Q17536_full_KJOH_table.qza",
  tree = "input_data/KJOH_16S_515F_926R_08052024_Q17536_full_KJOH_rooted-tree.qza",
  taxonomy = "input_data/KJOH_16S_515F_926R_08052024_Q17536_full_KJOH_silva-138-99-nb-classifier.qza",
    metadata = "input_data/KJOH_16S_515F_926R_08052024_Q17536_full_KJOH@mapping_formatted.txt"
  )

# write raw phyloseq object as .rds object
write_rds(physeq, "output_data/physeq.rds")

# check sample metadata and column types
sample_data(physeq)

```

```{bash}

#cp -R /Users/pvee/Wetsus_Projects/KJOH/KJOH_proj4_Q17535_XTO_monitoring/data_analysis/scripts ./scripts

```

```{r clean taxonomy, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

source("scripts/tax_clean_updated.R")

physeq_clean = tax_clean(physeq)

write_rds(physeq_clean, "output_data/physeq_clean_tax.rds")

```

#### Relative proportions of bacterial and non-bacterial sequences
```{r view non-bacterial sequences, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

psdata = physeq_clean
# clean non-informative and non-bacterial ASVs

# quick view on proportion non-bacterial sequences
psdata_nonBac =
psdata %>% 
  tax_glom(., "Phylum") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  select(Sample, Abundance, Kingdom, Phylum) %>% 
  filter(Kingdom != "Bacteria") %>% 
  group_by(Sample, Phylum) %>% 
  summarise(rel_abund = sum(Abundance)) 

psdata_nonBac %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Phylum)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")


# quick view on proportion Chloroplast and Mitochondrial sequences
psdata_mito_chlorop = 
physeq %>% 
  tax_glom(., "Family") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  filter(Kingdom == "Bacteria") %>% 
  select(Sample, Abundance, Kingdom, Phylum, Family) %>% 
  filter(grepl("Mitochondria|Chloroplast", Family)) %>% 
  group_by(Sample, Family) %>% 
  summarise(rel_abund = sum(Abundance))

psdata_mito_chlorop %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Family)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")

```

```{r clean on taxonomy, message=F, echo=F, eval=T, warning=F, include=T, cache=F}

# filter unwanted taxonomic groups out
psdata_tax_filtered =
psdata %>% # 652
  subset_taxa(.,grepl("Bacteria", Kingdom)) %>% # 0 ASVs lost because of other Kingdom than Bacteria
  subset_taxa(., Kingdom != "Unassigned") %>% # 0 ASVs lost
  subset_taxa(., !is.na(Phylum)) %>% # 0 ASVs lost 
  subset_taxa(., Phylum != "Phylum of Bacteria") %>% # 0 ASVs lost
  subset_taxa(., Family != "Mitochondria") %>% # 0 ASVs lost
  subset_taxa(., Class != "Chloroplast") %>% # 0 ASVs lost
  subset_taxa(., Order != "Chloroplast") # 0 ASVs lost


# remove non-informative taxa
old = ntaxa(psdata)
new = ntaxa(psdata_tax_filtered)

# number of ASVs removed:
old-new # 216 ASVs are removed

# check removed Mitochondrial ASV counts per sample
# psdata %>% 
#   subset_taxa(., Family == "Mitochondria") %>% 
#   psmelt %>% 
#   select(OTU, Sample, Family,Abundance) %>% 
#   mutate_if(is.character, as.factor) %>% 
#   group_by(Sample) %>%
# summarise(Abundance=sum(Abundance)) %>% 
# arrange(desc(Abundance), .by_group = T)

# save unfiltered psdata
psdata_unfiltered = psdata 
saveRDS(object = psdata_unfiltered, file = "output_data/psdata_unfiltered.rds") # save unfiltered data
psdata = psdata_tax_filtered # continue with filtered data as psdata

```

```{r clean phylogeny, message=F, echo=F, eval=T, warning=T, include=T, cache=F}

### resolve phylogenetic tree ###
physeq = physeq_clean

# evaluate tree topology
is.binary(phy_tree(psdata_tax_filtered)) # if FALSE --> polychotomy present (node has more than 2 tips)
#TRUE

# # if FALSE:
# # resolve polychotomous nodes
phy_tree_resolved <- multi2di(phy_tree(physeq))
is.binary(phy_tree_resolved)
# create new phy_tree
tree2 <- phy_tree_resolved

# subset_taxa(phy_tree_resolved, Kingdom ==  "Bacteria")

# merge new phy_tree object with sample_data and otu_table into new phyloseq object
physeq2 <- merge_phyloseq(otu_table(physeq), sample_data(physeq), tax_table(physeq), tree2)

# now filter only the taxa that are clean_filtered by taxonomic filtering
psdata = subset_taxa(physeq2, taxa_names(physeq2) %in% taxa_names(psdata_tax_filtered))

# save filted phyloseq object with resolved tree
saveRDS(object = psdata, file = "output_data/psdata_filtered_tree-resolved.rds")

```

```{r rarefaction curves, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# alpha rarefaction curve
source("scripts/ampvis2_internals.r")
source("scripts/amp_rankabundance.r")
source("scripts/amp_rarecurve.r")

# show plot
amp_rarecurve(psdata, color = "Description", legend.position = "bottomright", xlim = c(0, 120000))

# save plot
pdf(glue("figures/{proj}_rarefaction_curves.pdf"), useDingbats = F, width = 7, height = 5)
amp_rarecurve(psdata, color = "Description", xlim = c(0, 120000), legend.position = "bottomright")
dev.off()


```

```{r view mock and blank, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# extract DNA extraction blank # 216 ASVs
psdata_blank = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, sample_type  == "blank")) > 0,
    subset_samples(psdata, sample_type  == "blank")
    )

# extract mock community # 287 ASVs
psdata_mock = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, sample_type  == "mock")) > 0,
    subset_samples(psdata, sample_type  == "mock")
    )

saveRDS(object = psdata_mock, file = glue::glue("~/Wetsus_Projects/PVEE/Data_analyses/mock_communities_wetsus/mock_communities_wetsus/input_data/{proj}_pH_effect.rds"))

```

```{r remove_mock_and_nc, echo=F}

# keep samples only # 474 ASVs
psdata_samples = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, grepl("sample", sample_or_control))) >0,
    subset_samples(psdata, grepl("sample", sample_or_control))
    )

# recreate psdata with samples only
psdata = psdata_samples

```

```{r filter on abundance, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# relative abundance data
psdata_rel <- transform_sample_counts(psdata, fun = function(x) x/sum(x)) # psdata_rel taxa
ntaxa(psdata_rel)

# total sequence sum
total_sum = sum(sample_sums(psdata)) # 2588614 reads

# continue downstream analysis with abundance filter that retains ASVs with at least 0.1% of total read abundance -> keeps above 99% abundance when removing rare data
psdata_0.01pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0001, psdata) # 187 taxa (99.9% abundance)
psdata_0.05pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0005, psdata) # 114 taxa (99.8% abundance)
psdata_0.1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.001, psdata) # 91 taxa (99.7% abundance)
psdata_1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.01, psdata) # 34 taxa (97.8% abundance)

sum(sample_sums(psdata_0.01pct))/total_sum*100
sum(sample_sums(psdata_0.05pct))/total_sum*100
sum(sample_sums(psdata_0.1pct))/total_sum*100
sum(sample_sums(psdata_1pct))/total_sum*100

psdata_NoAbundanceFilter = psdata
psdata <- psdata_0.1pct # keeping 99.7% of reads in 91 ASVs

# difference in sample coverage (1.29)
max(sample_sums(psdata_0.1pct)/min(sample_sums(psdata_0.1pct)))

```

```{r prepare data for picrust2, eval=F}

# input data = psdata_0.05pct
biom_test =
psdata_0.1pct %>% 
  otu_table() %>% 
  data.frame(rownames = 1) %>% 
  as_tibble()

biomformat::write_biom(biom_test, glue("output_data/{proj}_otu_table_psdata_0.1pct.biom"))

# try running PICRUST2 as QIIME2 plugin
# DONE

```

#### Agglomeration of ASVs to Genus level groups for downstream analysis for both low N and high N sub-experiments
```{r agglomerate to genus level}

# psdata Genus 
psdata_Genus = tax_glom(
  transform_sample_counts(
    psdata, 
    function(x) x/sum(x)), "Genus")

psdata_Genus

# psmelts Genus
ps_Genus = psmelt(psdata_Genus) # n=57

```

#### Write output file usable as input for primer design tool (PDT)
```{r create PDT input Genus, echo=F, eval=T}

# create absolute genus physeq
# psdata Genus 
psdata_Genus_abs = tax_glom(psdata, "Genus")

long = 
psdata_Genus_abs %>% 
  psmelt() %>% 
  as_tibble()

# calculate genus-level taxa_sums across samples and calculate taxa prevalence %
wide_taxsums_prev = long %>% 
  select(Sample, Kingdom:Genus, Abundance) %>% 
  pivot_wider(names_from = "Sample", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(taxa_sums = sum(c_across(any_of(starts_with("KJOH")))),
         prevalence = mean(c_across(any_of(starts_with("KJOH"))) != 0, na.rm = TRUE) * 100) %>% 
  select(Kingdom:Genus, prevalence, taxa_sums, everything()) 

# make data long again
long_taxsums_prev = 
wide_taxsums_prev %>%
  pivot_longer(cols = any_of(starts_with("KJOH")), names_to = "Sample", values_to = "Abundance")

# calculate sample_sums for sample set
long_sampsums = 
wide_taxsums_prev %>% 
  pivot_longer(cols = any_of(starts_with("KJOH")), names_to = "Sample", values_to = "Abundance") %>% 
  select(Genus, Sample, Abundance) %>% 
  pivot_wider(names_from = "Genus", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(sample_sums = sum(c_across(2:ncol(.)))) %>% 
  select(Sample, sample_sums, everything()) %>% 
  pivot_longer(cols = c(3:ncol(.)), names_to = "Genus", values_to = "Abundance")

## PDT data formatted

# which prevalence and abundance thresholds in (%)
# vectors should be of same lengths corresponding to the combinations to make
prevalences = c(80, 90)
abundances = c(0.5, 0.1)

# loop over the data applying combinations of prevalences
for(prev in prevalences) {
  for(abun in abundances) {
    
inner_join(long_sampsums, long_taxsums_prev, by = c("Sample", "Genus")) %>% 
  select(-Abundance.y) %>% 
  rename(Abundance.x = "Abundance") %>% 
  mutate(perc_abund = Abundance/sample_sums*100) %>% 
  select(Sample, Genus, perc_abund, prevalence, "genus_sums" = taxa_sums, Family) %>% 
  group_by(Family, Genus, prevalence, genus_sums) %>% 
  summarise(mean_perc_abund = mean(perc_abund), max_perc_abund = max(perc_abund), min_perc_abund = min(perc_abund), .groups = "drop") %>% 
  filter(prevalence >= prev, mean_perc_abund >= abun) %>% 
  arrange(desc(max_perc_abund)) %>% 
  mutate(`0_Index` = row_number(),
         `1_Name` = if_else(grepl("Genus of ", Genus), Family, Genus),
         `2_Tax` = if_else(grepl("Genus of ", Genus), "family", "genus"),
         `3_mean_genus_perc` = mean_perc_abund,
         `4_perc_samples` = prevalence) %>% 
  select(`0_Index`, `1_Name`, `2_Tax`, `3_mean_genus_perc`, `4_perc_samples`, Family) %>% 
  arrange(desc(`2_Tax`)) %>% 
  write_csv(glue::glue("output_data/{proj}_PDT_input_Genus_prev{prev}_abund{abun}pct.csv"), col_names = T)
  }
}

cat("Input data for PDT saved as output table")

```


```{r alpha diversity rarefy}

source("scripts/avgrarefy.r")

# create subset frequency matrix
KJOH_matrix <- as(t(otu_table(psdata_Genus_abs)), "matrix")

# determine minimal sampling depth
min_sample <- min(sample_sums(psdata_Genus_abs)) # 192213

# rarefaction taking mean of 100 iterations
set.seed(711)
KJOH_matrix_rare192213_table = avgrarefy(x=KJOH_matrix, sample = min_sample, iterations = 100, seed = 711)

# create phyloseq object with rarefied data
psdata_KJOH_rare <- psdata_Genus
otu_rare = otu_table(data.frame(t(KJOH_matrix_rare192213_table)), taxa_are_rows = TRUE)
otu_table(psdata_KJOH_rare) <- otu_rare

```


#### Alpha Diversity
Analysis of genus level alpha diversity based on ASV-level rarefied data
```{r plot alpha diversity rare, message=F, echo=F, eval=T, warning=T, include=T, cache=F}

# calculate alpha diversity: richness & Shannon on Genus level
# input = 
psdata_KJOH_rare

# rarefied
alpha <- estimate_richness(psdata_KJOH_rare, measures = c("Observed", "Chao1", "Shannon", "Simpson"))
alpha$Sample <- row.names(alpha)

metadata = 
  sample_data(psdata_KJOH_rare) %>% 
  data.frame() %>% 
  as_tibble() %>% 
  mutate(Sample = sample_names(psdata_KJOH_rare)) %>% 
  select(Sample, everything()) 

# join alpha diversity with metadata
alpha2 <- inner_join(metadata, alpha, by = "Sample") #%>% 
  #mutate(biofilm = factor(biofilm, levels = c("Broth", "Biofilm")))

# plot alpha diversity
Chao1 <- 
alpha2 %>% 
  mutate(Sample_date = lubridate::dmy(Sample_date),
         Timeline = factor(Timeline, levels = c("start", "end"))) %>% 
ggplot(aes(x=as.factor(pH), y=Chao1)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  # geom_point(aes(fill = Timeline, color = Timeline),
  #                #position = position_jitter(width = 0.3),
  #                size = 2, shape = 21, show.legend = F) + 
  geom_col(aes(fill = Timeline, color = Timeline),
                 position = position_dodge(width = 0.7),
                 show.legend = T, width = 0.5) +
  scale_color_manual(values = c("steelblue", "darkred")) +
  scale_fill_manual(values = c("steelblue", "darkred")) +
  #scale_x_date(date_labels = "%b", date_breaks  ="1 month") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Genus Chao1"),
       x = "pH") +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  facet_grid(cols = vars(Set_up), scales = "free", space = "free")

print(Chao1)



# plot alpha diversity
Shannon <- 
alpha2 %>% 
  mutate(Sample_date = lubridate::dmy(Sample_date),
         Timeline = factor(Timeline, levels = c("start", "end"))) %>% 
ggplot(aes(x=as.factor(pH), y=Shannon)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  # geom_point(aes(fill = Timeline, color = Timeline),
  #                #position = position_jitter(width = 0.3),
  #                size = 2, shape = 21, show.legend = F) + 
  geom_col(aes(fill = Timeline, color = Timeline),
                 position = position_dodge(width = 0.7),
                 show.legend = T, width = 0.5) +
  scale_color_manual(values = c("steelblue", "darkred")) +
  scale_fill_manual(values = c("steelblue", "darkred")) +
  #scale_x_date(date_labels = "%b", date_breaks  ="1 month") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Genus Shannon H'"),
       x = "pH") +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  facet_grid(cols = vars(Set_up), scales = "free", space = "free")


print(Shannon)

# create plot
prow = 
  plot_grid(
  plot_grid(Chao1 + theme(legend.position = "none"), 
            Shannon + theme(legend.position = "none"),
                   align = "hv",
                   labels = c("A", "B"),
                   hjust = -1, 
                   nrow = 1),
  get_legend(Chao1), 
  nrow = 2 , 
  rel_heights = c(5, 1))


# show plot
prow

# save plot
ggsave(prow, filename = glue("figures/{proj}_plot_alpha_div.pdf"), width = 7, height = 5)
 
# save Shannon plot
ggsave(Shannon, filename = glue("figures/{proj}_plot_alpha_div.pdf"), width = 7, height = 4)

```

```{r alpha statistics}

# model_alpha = 
# alpha2 %>% 
#   mutate(across(everything(), ~str_replace_all(.x, ",", "."))) %>% 
#   mutate(across(everything(), ~type.convert(.x, as.is = TRUE))) %>% 
#   mutate(Sample_date = lubridate::dmy(Sample_date)) %>% 
#   with(lm(Shannon ~  pH + Sulfate + Total_Na)) 
# 
# model_alpha = update(model_alpha)
# # anova result
# anova(model_alpha)
# 
# # test differential effects of time on Shannon between substrates and sludge
# emmeans_shannon = emmeans::emmeans(model_alpha, pairwise ~ time | substrate + sludge_source, p.adjust="Holm")
# 
# # save anova contrasts
# emmeans_shannon$contrasts %>% 
#   as_tibble() %>% 
#   write_csv(glue::glue("output_data/{proj}_anova_result_Shannon_Tukey_pvals.csv"))

```

```{r write genus tables, echo=F}

# paper 1 only
### write rel abund data plus other tax info
ps_Genus %>% 
  # filter(!grepl("mock", combined), # remove mock communities
  #        Sample_mock != "blank") %>% # remove NC blanks
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, OTU, Abundance, Kingdom:Genus, everything()) %>% 
  group_by_at(vars(Sample, Kingdom:Genus)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(-rel_abund)) %>% 
  summarise(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  select(Kingdom:Genus, Sample, mean_rel_abund) %>% 
  pivot_wider(names_from = Sample, values_from = mean_rel_abund, values_fill = 0) %>% 
  group_by_at(vars(Kingdom:Genus)) %>% 
  rowwise() %>% 
  mutate(mean = mean(c_across(where(is.numeric)))) %>% 
  arrange(desc(mean)) %>% 
  select(-mean) %>% 
  ungroup() %>% 
  mutate(across(where(is.numeric), ~round(.x, digits = 2))) %>% 
  select(Kingdom:Genus, everything()) %>% 
  write_csv(glue::glue("output_data/{proj}_relative_Genus_abundances_sorted_by_overall_mean.csv", col_names = T, append = F))

```

#### Barplots of genus level taxonomy
```{r plot_bar all genera, echo=FALSE, fig.width=8, fig.height=5}

#ps_Genus

# count n genera n = 18
ps_Genus %>% 
  as_tibble() %>% 
  mutate(Genus = factor(Genus)) %>% 
  pull(Genus) %>% 
  levels() %>% 
  length()

# Define a function to find the first non-matching column

# Use rowwise and mutate to apply the function
ps_Genus_taxlabel = ps_Genus %>% 
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, Abundance, Sample_date, exp_id, Set_up, Timeline, pH:Genus) %>%
  rowwise() %>%
  mutate(tax_label = case_when(
    grepl("\\d", Genus) ~ {
      first_non_numeric <- case_when(
        !grepl("\\d", Family) ~ paste(Family, Genus, sep = " "),
        !grepl("\\d", Order) ~ paste(Order, Genus, sep = " "),
        !grepl("\\d", Class) ~ paste(Class, Genus, sep = " "),
        !grepl("\\d", Phylum) ~ paste(Phylum, Genus, sep = " "),
        !grepl("\\d", Kingdom) ~ paste(Kingdom, Genus, sep = " "),
        TRUE ~ NA_character_
      )
      first_non_numeric
    },
    TRUE ~ Genus
  ))

Genus_abundances <- ps_Genus_taxlabel %>% 
  mutate(Sample_nr = str_sub(Sample, start = 1, end = 2)) %>% 
  group_by_at(vars(Sample, Sample_date, exp_id, Set_up, Timeline, tax_label)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>%
  group_by(Sample) %>% 
  mutate(mean_rel_abund = rel_abund/sum(rel_abund)*100) %>% 
  ungroup() %>% 
  group_by(Sample, exp_id, Set_up, Timeline, tax_label) %>% 
  mutate(tax_label = str_replace(tax_label, "(.*)_unclassified", "Unclassified *\\1*"),
         tax_label = str_replace(tax_label, "^(\\S*)$", "*\\1*"))

# Reduce genera to lower number (at least 3%, or else called other)
Genus_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 3, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

# define a color vector using brewer.pal and colorset
colvec = c(brewer.pal(8,"Dark2"),brewer.pal(8,"Set2"),brewer.pal(8,"Accent"),brewer.pal(8,"Paired"),brewer.pal(12,"Set3"),brewer.pal(9,"Set1"), brewer.pal(9,"YlGnBu"), colorset)

# name the taxon colors
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  group_by(tax_label) %>%
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            median = median(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, median, .desc = T)) %>% 
  arrange(desc(median)) %>% pull(tax_label) %>% unique() -> names(colvec)

colvec["Other max.<3%"] <- "#D3D3D3"

# reshape plot data
plot_data <-
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  # mutate(Sample = if_else(grepl(".*[^0-9][0-9]{2}$", Sample),
  #                          Sample,
  #                          str_c(str_sub(Sample, end = -2),"0", str_sub(Sample, -1)))) %>% 
  group_by(Sample, Sample_date, exp_id, Set_up, Timeline, tax_label) %>% 
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            mean = min(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, mean, .desc = T),
         Timeline = factor(Timeline, levels = c("start", "end"))) 


# strip labels
setup_label = c("Dual set-up", "Single set-up")
names(setup_label) = c("dual", "single")
pH_label = c("pH 8.5", "pH 9", "pH 9.3")
names(pH_label) = c("8.5", "9", "9.3")

# plot the data per sample
plot_Genus = 
  plot_data %>% 
  inner_join(., metadata %>% select(Sample, pH) %>% filter(Sample %in% .$Sample), by = "Sample") %>% 
  mutate(Sample_date = dmy(Sample_date),
         pH_factor = factor(pH)) %>% 
ggplot(aes(x=Timeline, y = mean_rel_abund, fill = tax_label)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colvec) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0), breaks = seq(0, 100, 10)) +
  #scale_x_date(date_labels = "%d-%b-%Y", date_breaks = "1 month") +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom",
        strip.background = element_rect(colour = "white"),
        strip.text = element_text(face = "bold")) +
  guides(fill = guide_legend(nrow=7)) +
  facet_nested(~Set_up + pH_factor + Timeline, 
               labeller = labeller(Set_up = setup_label,
                                   pH_factor = pH_label),
               scales = "free", 
               nest_line = element_line(lineend = "round"), 
               resect = unit(3, "mm"))
  

# print plot  
gg_Genus = plot_Genus
gg_Genus

# save plots
ggsave(plot = gg_Genus, filename = glue("figures/{proj}_barplot_all_Genus_by_date.pdf"), width = 8, height = 6)

```



```{r geom_tile, echo = F, fig.width=10, fig.height=4.5}

# remarks order for x axis
#order = c(1,19,16,6,5,9,17,7,8,4,3,18,2,10,11,12,13,14,15)

taxon_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 1, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

heatmap_data_long = 
Genus_abundances %>% 
  group_by(Sample, Sample_date, Set_up, Timeline, tax_label) %>% 
  summarise(mean_rel_abund = mean(mean_rel_abund), .groups = "drop") %>% 
inner_join(., Genus_pool, by="tax_label") %>% 
  group_by(Sample, Sample_date, Set_up, Timeline, tax_label) %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  select(Sample, Sample_date,Set_up, Timeline, tax_label, mean_rel_abund, mean, pool) %>% 
  group_by(Sample, Sample_date, Set_up, Timeline, tax_label) %>% 
  summarise(sum_rel_abund = sum(mean_rel_abund)) %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, sum_rel_abund, .desc = F)) %>%
  ungroup()

# determine order of taxa on y axis from mean high to low
tax_order = 
heatmap_data_long %>% 
  mutate(tax_label = as.character(tax_label)) %>% 
  group_by(tax_label) %>% 
  summarise(mean = mean(sum_rel_abund), .groups = "drop") %>% 
  arrange(desc(mean)) %>% pull(tax_label)

# create wide format data
heatmap_data = 
heatmap_data_long %>% 
select(Sample, Sample_date, Set_up, Timeline, tax_label, sum_rel_abund) %>% 
  pivot_wider(names_from = tax_label, 
              values_from = sum_rel_abund) 

# make heatmap
m <- as.matrix(heatmap_data[, 5:ncol(heatmap_data)])
clust <- hclust(dist(t(m)))


                            
### TODO WITH THE INCREASE/DECREASE/STABLE CATEGORIZATION OF THE HEATMAP
# plot heatmap
heatmap_genus_3pct = 
heatmap_data_long %>% 
  inner_join(., metadata %>% select(Sample, pH) %>% filter(Sample %in% .$Sample), by = "Sample") %>% 
  mutate(Sample_date = dmy(Sample_date),
         Sample_date = as.character(Sample_date), 
         pH_factor = factor(pH), 
         Timeline = factor(Timeline, levels = c("start", "end"))) %>% 
ggplot(aes(Sample_date, tax_label)) +
  geom_tile(aes(fill = sum_rel_abund), color = NA) +
  labs(x = NULL, y=NULL) +
  scale_y_discrete(limits = colnames(m)[rev(clust$order)]) +
  scale_fill_gradient(low = "white", high = "darkred",
                      name = "Relative\nAbundance (%)\n(\u00B7 0%)") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_blank(), #element_text(size = 8, angle = 90, hjust = 1, vjust = -1),
        strip.text.x = element_text(angle = 0),
        strip.text.y = element_text(angle = 0),
        axis.ticks.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_markdown(size = 8),
        plot.background = element_blank(),
        strip.placement = "outside",
        strip.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        panel.spacing.x = unit(c(-1.5, 2, -1.5, 2, -1, 5, -1.5, 2, -1.5, 2, -1.5), "mm")
        ) +
  geom_text(
  aes(label = ifelse(sum_rel_abund >1 & sum_rel_abund < 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey50", size =2.5) +
  geom_text(
  aes(label = ifelse(sum_rel_abund == 0, ".", "")),
  color = "grey80", size = 5, position = position_nudge(y=0.5)) +
  geom_text(
  aes(label = ifelse(sum_rel_abund > 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey90", size = 2.5) +
  facet_nested(~Set_up + pH_factor + Timeline, 
               labeller = labeller(Set_up = setup_label,
                                   pH_factor = pH_label),
               scales = "free", 
               nest_line = element_line(lineend = "round"), 
               resect = unit(3, "mm"))
  
  

 print(heatmap_genus_3pct)

ggsave2(plot = heatmap_genus_3pct + 
                ggtitle(label = "Genus heatmap - XTO monitoring"), 
        filename = glue("figures/{proj}_heatmap_genus.pdf"), width = 10, height = 4.5)

# save plot data with max.<3% others
heatmap_data_long %>% 
  mutate(tax_label = fct_relevel(tax_label, rev(tax_order))) %>% 
  select(Sample, tax_label, sum_rel_abund) %>%
  pivot_wider(names_from = "Sample", values_from = "sum_rel_abund", values_fill = 0) %>%
write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table.xlsx"), col_names = T)

# save plot data without other category
ps_Genus %>% 
  as_tibble() %>% 
  select(Sample, Sample_date, Abundance, Kingdom:Genus) %>% 
  group_by_at(vars(Sample, Sample_date, Kingdom:Genus)) %>% 
  summarize(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(Sample_date, Kingdom:Genus)) %>% 
  summarize(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  mutate(Genus = str_replace(Genus, "(.*)_unclassified", "Unclassified *\\1*"),
         Genus = str_replace(Genus, "^(\\S*)$", "*\\1*")) %>% 
  pivot_wider(names_from = "Sample_date", values_from = "mean_rel_abund", values_fill = 0) %>%
  write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table_all.xlsx"), col_names = T) 
  
```

### Beta diversity

```{r total beta diversity rel_abund, message=F, fig.width=12, fig.height=5, echo=F, eval=T, warning=F, include=T, cache=F}

# recode days as categorical
colnames(phyloseq::sample_data(psdata_Genus))

### Beta diversity analysis first
# input data  = psdata
psdata_Genus_rel = transform_sample_counts(psdata, function(x) x/sum(x))
psdata_Genus_clr = microbiome::transform(psdata_Genus, transform = "clr")

# reformat date variable
phyloseq::sample_data(psdata_Genus_rel)$Sample_date = lubridate::dmy(phyloseq::sample_data(psdata_Genus_rel)$Sample_date)

# calculate numerically days between dates for metadata
phyloseq::sample_data(psdata_Genus_rel)$days = as.numeric(difftime(phyloseq::sample_data(psdata_Genus_rel)$Sample_date, min(phyloseq::sample_data(psdata_Genus_rel)$Sample_date), units = "days") + 1)

# create pH factor in metadata
phyloseq::sample_data(psdata_Genus_rel)$pH_factor = as.factor(phyloseq::sample_data(psdata_Genus_rel)$pH) 

# ordination
PCoA_BC <- ordinate(psdata_Genus_rel, method = "PCoA", distance = "bray")
PCoA_Jac <- ordinate(psdata_Genus_rel, method = "PCoA", distance = "jaccard")
PCoA_uu <- ordinate(psdata_Genus_rel, method = "PCoA", distance = "uunifrac")
PCoA_wu <- ordinate(psdata_Genus_rel, method = "PCoA", distance = "wunifrac")

nmds_BC <- ordinate(psdata_Genus_rel, method = "NMDS", distance = "bray")
nmds_Jac <- ordinate(psdata_Genus_rel, method = "NMDS", distance = "jaccard")
nmds_uu <- ordinate(psdata_Genus_rel, method = "NMDS", distance = "uunifrac")
nmds_wu <- ordinate(psdata_Genus_rel, method = "NMDS", distance = "wunifrac")

plot_PCoA_BC <- 
  plot_ordination(physeq = psdata_Genus_rel, 
                  ordination = PCoA_BC, 
                  type = "samples", 
                  axes = c(1,2), 
                  color = "pH_factor", 
                  shape = "Timeline") + 
  #ggrepel::geom_text_repel(aes(label = days), size = 2) +
  scale_color_manual(values = brewer.pal(6, "Set2")) +
  geom_point(size=2) +
  ggtitle("Bray-Curtis\n(presence + abundance)") +
theme(
  plot.title = element_text(hjust = 0.5, vjust = 3),
  panel.background = element_rect(fill='transparent'),
  panel.grid = element_line(colour = "grey90"),
  strip.text = element_text(face = "bold"),
  panel.border = element_rect(colour = "black", fill=NA, linewidth = 0.3),
  axis.line.x = element_line(color="black", size = 0.3),
  axis.line.y = element_line(color="black", size = 0.3)) +
  facet_wrap(~Set_up)
  
  
plot_PCoA_Jac <- 
    plot_ordination(physeq = psdata_Genus_rel, 
                  ordination = PCoA_Jac, 
                  type = "samples", 
                  axes = c(1,2), 
                  color = "pH_factor", 
                  shape = "Timeline") + 
  #ggrepel::geom_text_repel(aes(label = days), size = 2) +
  scale_color_manual(values = brewer.pal(6, "Set2")) +
  geom_point(size=2) +
  ggtitle("Jaccard\n(presence only)") +
theme(
  plot.title = element_text(hjust = 0.5, vjust = 3),
  panel.background = element_rect(fill='transparent'),
  panel.grid = element_line(colour = "grey90"),
  strip.text = element_text(face = "bold"),
  panel.border = element_rect(colour = "black", fill=NA, linewidth = 0.3),
  axis.line.x = element_line(color="black", size = 0.3),
  axis.line.y = element_line(color="black", size = 0.3)) +
  facet_wrap(~Set_up)

plot_PCoA_wu <- 
  plot_ordination(physeq = psdata_Genus_rel, 
                  ordination = PCoA_wu, 
                  type = "samples", 
                  axes = c(1,2), 
                  color = "pH_factor",
                  shape = "Timeline") + 
  #ggrepel::geom_text_repel(aes(label = days), size = 2) +
  scale_color_manual(values = brewer.pal(6, "Set2")) +
  geom_point(size=2) +
theme(
  plot.title = element_text(hjust = 0.5, vjust = 3),
  panel.background = element_rect(fill='transparent'),
  panel.grid = element_line(colour = "grey90"),
  strip.text = element_text(face = "bold"),
  panel.border = element_rect(colour = "black", fill=NA, linewidth = 0.3),
  axis.line.x = element_line(color="black", size = 0.3),
  axis.line.y = element_line(color="black", size = 0.3)) +
  facet_wrap(~Set_up)


# Combine the plots with shared legend
combined_plots <- patchwork::wrap_plots(
  plot_PCoA_Jac + ggtitle("Jaccard distance\n(Genus presence)"),
  plot_PCoA_wu + ggtitle("weighted UniFrac\n(lineage presence and abundance)"),
  plot_PCoA_BC + ggtitle("Bray-Curtis dissimilarity\n(Genus presence and abundance)")) +
  patchwork::plot_layout(guides = "collect", ncol = 3) 

# Display the combined plot
combined_plots

ggsave(plot = combined_plots, filename = glue("figures/{proj}_plot_beta_diversity_all_rel.pdf"), width = 12, height =4.5)


```

